src/Database/Esqueleto/Internal/Internal.hs

```diff
@@ -1759,7 +1834,7 @@ type SqlEntity ent = (PersistEntity ent, PersistEntityBackend ent ~ SqlBackend)
 
 -- | Side data written by 'SqlQuery'.
 data SideData = SideData { sdDistinctClause :: !DistinctClause
-                         , sdFromClause     :: ![FromClause]
+                         , sdFromClause     :: !FromClause
```

Comment:
So morally this *was* more like:

```
sdFromClause :: (FromClause, [OnClause])
```

and then post-processing would fill in the `Nothing` values with `Just OnClause`.

`FromClause` is a binary tree, and binary trees are monoids already. So I think we might be able to get this working without introducing the `Partial` type, possibly by altering the usage API somewhat.

Reply:
that was the understanding that I had until I realized that if you have two FromClause in that list that aren't OnClause it will join them with a ",". I didn't want to change this behavior in case there was some use.

Reply:
Yeah, that's usually useful for writing a cross join:

```sql
SELECT * FROM user, blog_post;
```

will just do a cartesian product. And so in Haskell we can write:

```haskell
select $ 
from $ \(person, blogPost) ->
pure (person, blogPost)
```

to do the same thing. Or replace the comma with `CrossJoin` to be more explicit. These clauses don't require na `on` clause because there's no filter step.